library/src/main/java/org/apache/apex/malhar/lib/window/impl/WindowedJoinOperatorFeatures.java

```diff
@@ -0,0 +1,183 @@
+    long minWatermark = Math.min(latestWatermark1, latestWatermark2);
+    operator.setCurrentWatermark(minWatermark);
+  }
+
+  private void setCurrentWatermark(long timestamp)
+  {
+    operator.setCurrentWatermark(Math.min(latestWatermark1, latestWatermark2));
+  }
+
+  /**
+   *
+   * @param <InputT1>
+   * @param <InputT2>
+   * @param <AccumT>
+   * @param <AccumulationT>
+   * @param <DataStorageT>
+   */
+  public static class Plain<InputT1, InputT2, AccumT, AccumulationT extends JoinAccumulation<InputT1, InputT2, AccumT, ?>, DataStorageT extends WindowedStorage.WindowedPlainStorage<AccumT>>
+      extends WindowedJoinOperatorFeatures<InputT1, InputT2, AccumT, AccumulationT, DataStorageT>
+  {
+    private Plain()
+    {
+      // for kryo
+    }
+
+    public Plain(AbstractWindowedOperator<InputT1, ?, DataStorageT, ?, AccumulationT> operator)
+    {
+      super(operator);
+    }
+
+    protected <T> void accumulateTuple(Tuple.WindowedTuple<T> tuple, AccumFunction<T>
+        accumFn)
+    {
+      for (Window window : tuple.getWindows()) {
+        // process each window
+        AccumT accum = operator.getDataStorage().get(window);
+        if (accum == null) {
+          accum = operator.getAccumulation().defaultAccumulatedValue();
+        }
+        operator.getDataStorage().put(window, accumFn.accumulate(accum, tuple.getValue()));
+      }
+    }
+
+    @Override
+    public void accumulateTuple1(Tuple.WindowedTuple<InputT1> tuple)
+    {
+      accumulateTuple(tuple, new AccumFunction<InputT1>()
+      {
+        @Override
+        AccumT accumulate(AccumT accum, InputT1 value)
+        {
+          return operator.getAccumulation().accumulate(accum, value);
+        }
+      });
+    }
+
+    @Override
+    public void accumulateTuple2(Tuple.WindowedTuple<InputT2> tuple)
+    {
+      accumulateTuple(tuple, new AccumFunction<InputT2>()
+      {
+        @Override
+        AccumT accumulate(AccumT accum, InputT2 value)
+        {
+          return operator.getAccumulation().accumulate2(accum, value);
+        }
+      });
+    }
+  }
+
+  /**
+   *
+   * @param <KeyT>
+   * @param <InputT1>
+   * @param <InputT2>
+   * @param <AccumT>
+   * @param <AccumulationT>
+   * @param <DataStorageT>
+   */
+  public static class Keyed<KeyT, InputT1, InputT2, AccumT, AccumulationT extends JoinAccumulation<InputT1, InputT2, AccumT, ?>, DataStorageT extends WindowedStorage.WindowedKeyedStorage<KeyT, AccumT>>
+      extends WindowedJoinOperatorFeatures<KeyValPair<KeyT, InputT1>, KeyValPair<KeyT, InputT2>, AccumT, AccumulationT, DataStorageT>
+  {
+    private Keyed()
+    {
+      // for kryo
+    }
+
+    public Keyed(AbstractWindowedOperator<KeyValPair<KeyT, InputT1>, ?, DataStorageT, ?, AccumulationT> operator)
+    {
+      super(operator);
+    }
+
+    protected <T extends KeyValPair> void accumulateTuple(Tuple.WindowedTuple<T> tuple, AccumFunction<T> accumFn)
+    {
+      final KeyValPair<KeyT, T> kvData = tuple.getValue();
```

Comment:
Sorry I didn't explain clearly. Here the kvData is the value of the tuple which is a KeyValPair, this should have no issue. But the type of the value of kvData is T, which extends KeyValPair. What I meant in my previous question is why the value of kvData is another KeyValPair?